### PR TITLE
[CWS] add security profile sliding window

### PR DESCRIPTION
**Pull request by spikat**
### What does this PR do?

This PR adds a life-cycle for security profiles. The selector is no longer focused on image_name + image_tag: now the selector is only image_name and each profiles could now contains several image_tag versions.

The maximum number of versions for a profile can be configured via `runtime_security_config.security_profile.max_image_tags` (10 by default).
Once the limit is reached, the last seen version is evicted.

The first version would come from a security dump, then further versions will be directly added by the profile manager.

New set of tests has been added to validate the changes:
```
TestSecurityProfileLifeCycleExecs/life-cycle-v1-learning-new-process
TestSecurityProfileLifeCycleExecs/life-cycle-v1-stable-process-anomaly
TestSecurityProfileLifeCycleExecs/life-cycle-v2-learning-new-process-anomaly
TestSecurityProfileLifeCycleExecs/life-cycle-v2-learning-v1-process
TestSecurityProfileLifeCycleExecs/life-cycle-v1-unstable-new-process
TestSecurityProfileLifeCycleDNS/life-cycle-v1-learning-new-dns
TestSecurityProfileLifeCycleDNS/life-cycle-v1-stable-dns-anomaly
TestSecurityProfileLifeCycleDNS/life-cycle-v2-learning-new-dns-anomaly
TestSecurityProfileLifeCycleDNS/life-cycle-v2-learning-v1-dns
TestSecurityProfileLifeCycleDNS/life-cycle-v1-unstable-new-dns
TestSecurityProfileLifeCycleEvictitonProcess/life-cycle-eviction-process-v1-learning-new-process
TestSecurityProfileLifeCycleEvictitonProcess/life-cycle-eviction-process-v1-stable-process-anomaly
TestSecurityProfileLifeCycleEvictitonProcess/life-cycle-eviction-process-v2-learning-new-process-anomaly
TestSecurityProfileLifeCycleEvictitonProcess/life-cycle-eviction-process-check-v1-evicted
TestSecurityProfileLifeCycleEvictitonProcess/life-cycle-eviction-process-v1-process-anomaly
TestSecurityProfileLifeCycleEvictitonDNS/life-cycle-eviction-dns-v1-learning-new-process
TestSecurityProfileLifeCycleEvictitonDNS/life-cycle-eviction-dns-v1-stable-process-anomaly
TestSecurityProfileLifeCycleEvictitonDNS/life-cycle-eviction-dns-v2-learning-new-process-anomaly
TestSecurityProfileLifeCycleEvictitonDNS/life-cycle-eviction-dns-check-v1-evicted
TestSecurityProfileLifeCycleEvictitonDNS/life-cycle-eviction-dns-v1-process-anomaly
TestSecurityProfileLifeCycleEvictitonProcessUnstable/life-cycle-eviction-process-unstable-v1-learning-new-process
TestSecurityProfileLifeCycleEvictitonProcessUnstable/life-cycle-eviction-process-unstable-v1-unstable
TestSecurityProfileLifeCycleEvictitonProcessUnstable/life-cycle-eviction-process-unstable-v2-learning
TestSecurityProfileLifeCycleEvictitonProcessUnstable/life-cycle-eviction-process-unstable-v3-learning
TestSecurityProfileLifeCycleEvictitonProcessUnstable/life-cycle-eviction-process-unstable-v3-process-anomaly
```

There is also new metrics to follow the new sec profile behaviors:
`.security_profile.evicted_versions` will be send each time a profile version has been evicted (when reaching the max number of concurrent versions).
`.security_profile.versions` will reflect the number of version for each profile

### Motivation

This will allow to raise coverage on short lived containers, allowing to learn from a previous version instead of starting from blank each time.

### Additional Notes

We'll have to pay attention on how it behaves on some workloads where the image_tag is not used to version the image (we know that some uses the same image with multiple purposes, specifying it through image_tag).

### Describe how to test/QA your changes

Run the tests is a first way. Play with images and change image tag each time, have a look on profile states, try to generate anomalies.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by YoannGh — COMMENTED** [review]
I think the `manager.OnCGroupDeletedEvent` function should use the profile selector (`"*"`) instead of the one from the cgroup model parameter. Otherwise the manager fails to lookup the corresponding profile.

**Comment by spikat** [comment]
> I think the `manager.OnCGroupDeletedEvent` function should use the profile selector (`"*"`) instead of the one from the cgroup model parameter. Otherwise the manager fails to lookup the corresponding profile.

Thanks, good catch! Fixed in the last commit :)

File: pkg/security/security_profile/activity_tree/process_node.go [hunk 1]
```diff
@@ -319,9 +333,103 @@ func (pn *ProcessNode) InsertBindEvent(evt *model.Event, generationType NodeGene
...
+// also, recompute the list of dnsnames and syscalls
+func (pn *ProcessNode) EvictImageTag(imageTag string, DNSNames *utils.StringKeys, SyscallsMask map[int]int) bool {
+	if imageTag == "" {
+		return false
+	}
+
+	if !slices.Contains(pn.ImageTags, imageTag) {
+		return false // this node don't have the tag, and all his childs/files/dns/etc shouldn't have neither
+	}
+	pn.ImageTags = removeImageTagFromList(pn.ImageTags, imageTag)
```
Comment (safchain) [inline pkg/security/security_profile/activity_tree/process_node.go:396]: What about a removeImageTagFromList with the imageTag == "" &&  that will return if it succeeded or not

File: pkg/security/rconfig/profiles.go [hunk 2]
```diff
@@ -66,18 +66,22 @@ func (r *RCProfileProvider) rcProfilesUpdateCallback(configs map[string]state.Ra
 			continue
 		}
 
-		imageName := utils.GetTagValue("image_name", profile.Tags)
-		imageTag := utils.GetTagValue("image_tag", profile.Tags)
-
+		// get image name from tags of the first version
+		if len(profile.ProfileContexts) == 0 {
```
Comment (safchain) [inline pkg/security/rconfig/profiles.go:70]: What about having the image name at the top level and just a list of image_tags in the profilecontext tags ?

File: pkg/security/security_profile/profile/profile.go [hunk 3]
```diff
@@ -157,21 +173,48 @@ func (p *SecurityProfile) SendStats(client statsd.ClientInterface) error {
...
 		ProfileCookie: p.profileCookie,
-		Version:       p.Version,
 		Metadata: &api.MetadataMessage{
 			Name: p.Metadata.Name,
 		},
-		Tags: p.Tags,
+		ProfileGlobalState: p.GetGlobalState().String(),
+		ProfileContexts:    make(map[string]*api.ProfileContextMessage),
+	}
+	p.versionContextsLock.Lock()
```
Comment (YoannGh) [inline pkg/security/security_profile/profile/profile.go:200]: Should this lock be moved at beginning of the function (before the imageTags string is constructed)?

File: pkg/security/rconfig/profiles.go (hunk 2 above)
Comment (spikat) [reply pkg/security/rconfig/profiles.go:70]: I agree. The change is already present on the `yoanngh/persist-profiles` branch (adding the profile selector on profile top level). I'll ensure to include this change on the second PR.

File: pkg/security/security_profile/activity_tree/process_node.go (hunk 1 above)
Comment (spikat) [reply pkg/security/security_profile/activity_tree/process_node.go:396]: Good idea! I'll do it

File: pkg/security/security_profile/profile/profile.go (hunk 3 above)
Comment (spikat) [reply pkg/security/security_profile/profile/profile.go]: Suggestion applied on last commit, thank you!

File: pkg/security/security_profile/activity_tree/process_node.go [hunk 4]
```diff
@@ -319,9 +333,103 @@ func (pn *ProcessNode) InsertBindEvent(evt *model.Event, generationType NodeGene
...
+	}
+}
+
+// TagAllNodes tags this process, its files/dns/socks and childrens with the given image tag
+func (pn *ProcessNode) TagAllNodes(imageTag string) {
+	if imageTag == "" {
+		return
+	}
+
+	if !slices.Contains(pn.ImageTags, imageTag) {
```
[the remark this answers is not in the document]
Comment (spikat) [reply pkg/security/security_profile/activity_tree/process_node.go]: Indeed this pattern is repeated multiple times. I've updated the code by calling a new func doing all of that (without having a special type). Tell me what you think of it

**Review by safchain — APPROVED** [review]

**Review by YoannGh — APPROVED** [review]

**Pull request merged**